### PR TITLE
Gadget for opcode calldatasize

### PR DESCRIPTION
**Pull request by scroll-dev**
Spec: https://github.com/appliedzkp/zkevm-specs/pull/96

### CODE REVIEW

**Comment by ChihChengLiang** [comment]
@roynalnaruto Can we rebase this?

File: bus-mapping/src/evm/opcodes.rs
Comment (ed255) [inline bus-mapping/src/evm/opcodes.rs:103]: I think this can't be just a `StackOnlyOpcode` because it does a `CallContextOp` READ to the `rw_table`, similar to this (from https://github.com/appliedzkp/zkevm-circuits/pull/287): https://github.com/appliedzkp/zkevm-circuits/blob/0e7c1791f849fcb76d6f3367298313e0e5854c4f/bus-mapping/src/evm/opcodes/caller.rs#L21-L26

The `CallContextOp` is pending to be introduced in #292

**Review by ed255 — CHANGES_REQUESTED** [review]

File: bus-mapping/src/evm/opcodes.rs
Comment (roynalnaruto) [reply bus-mapping/src/evm/opcodes.rs:103]: Yes that's true, so let's wait until #292 is merged. Thanks!

File: zkevm-circuits/src/evm_circuit/execution/calldatasize.rs
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:44]: I believe that the CallDataLength from the CallContext is the actual value, but in the stack it must be RLC encoded:
https://github.com/appliedzkp/zkevm-specs/blob/feb9d0e9a42cd6ef3a62ae0d3e87cd9224100bb6/src/zkevm_specs/evm/execution/calldatasize.py#L13-L22

**Review by ed255 — CHANGES_REQUESTED** [review]

File: bus-mapping/src/evm/opcodes.rs
Comment (ed255) [reply bus-mapping/src/evm/opcodes.rs:103]: #292 has been merged :)

Comment (roynalnaruto) [reply bus-mapping/src/evm/opcodes.rs:103]: Thanks, will update my PR tomorrow 👍

File: zkevm-circuits/src/evm_circuit/execution/calldatasize.rs
Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:38]: @ed255 I've made changes to resolve the above comments and discussions.

Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:38]: I think this implementation is still different than the spec.  Currently the value used for the stack WRITE is not RLC encoded.

Take a look for example at the COINBASE implementation: https://github.com/appliedzkp/zkevm-circuits/blob/b1e368ec7812ca7db9a016b780d586e247f634a3/zkevm-circuits/src/evm_circuit/execution/coinbase.rs#L32  The coinbase value is an ethereum address so it fits in the field.  In the RwTable.CallContext it's raw, and then in the RwTable.Stack it's RLC encoded.

File: zkevm-circuits/src/evm_circuit/execution/calldatasize.rs
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:126]: Following the spec, this would be RLC encoded. See https://github.com/appliedzkp/zkevm-specs/blob/feb9d0e9a42cd6ef3a62ae0d3e87cd9224100bb6/tests/evm/test_calldatasize.py#L39

Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:126]: @ed255 I've used the `query_rlc` for calldatasize in the latest commit. Although this `value` field in the `Rw::Stack` is later converted into RLC-encoded value here:
https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/witness.rs#L631

Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:126]: Oh ok! Got it!  Sorry for the confusion.

Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:126]: Although the points you made earlier (regarding the fact that `cb.stack_pop` for the `call_data_size` value) were absolutely correct, and I've made the necessary changes for that.

Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:126]: In the current state of this PR the opcode expects that in the RwTable.Stack and RwTable.CallContext.CallDataLength the values are RLC encoded.  This is because you're using `call_data_size.expr()`.  The correct implementation should be that in RwTable.Stack it's RLC encoded, but in RwTable.CallContext.CallDataLength it's the actual value.  The actual value can be obtained by using `from_bytes::expr(&call_data_size.cells)`.  Check the coinbase opcode implementation for reference.

Now, In the current state of this PR the tests are passing, so I was trying to figure out why.  And it turns out that for values between `0` and `255`, `RLC(x) == x`, which is unfortunate, because it means that when we use small values in tests we can't catch bugs when comparing RLC encoded to non-RLC encoded values.  This is because the RLC encoding does `byte0 + byte1 * rnd + byte2 * rnd**2 + ...`.  So if only byte0 is filled, the rnd is not used. 

To validate that the current implementation is not correct, please add a test case for a value of `300`, and you'll see that it fails.

Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:126]: My bad, I think I should have also added a test case with a random calldatasize with upper bound on bit length. And this bug would have been caught. Thanks so much :) Will fix it today.

Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/calldatasize.rs:136]: @ed255 The issue should now be resolved. I've also added another constant `N_BYTES_CALLDATASIZE` which is the same as `u64` (or 8 bytes), since that's what [`go-ethereum` limits to](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L294-L297) as well.

**Review by ed255 — APPROVED** [review]
LGTM! Great work :)

**Comment by ed255** [comment]
@roynalnaruto this PR needs one last rebase with main and it will be ready to be merged :)

**Comment by roynalnaruto** [comment]
> @roynalnaruto this PR needs one last rebase with main and it will be ready to be merged :)

Done! 👍

**Pull request merged**
